src/main/java/htsjdk/beta/codecs/reads/bam/BAMEncoder.java

```diff
@@ -0,0 +1,37 @@
+package htsjdk.beta.codecs.reads.bam;
+
+import htsjdk.beta.plugin.bundle.Bundle;
+import htsjdk.beta.plugin.bundle.BundleResourceType;
+import htsjdk.beta.plugin.reads.ReadsEncoderOptions;
+import htsjdk.beta.plugin.reads.ReadsFormats;
+import htsjdk.beta.plugin.reads.ReadsEncoder;
+
+/**
+ * Base class for {@link htsjdk.beta.plugin.bundle.BundleResourceType#READS_BAM} encoders.
+ */
+public abstract class BAMEncoder implements ReadsEncoder {
+    protected final Bundle outputBundle;
+    protected final ReadsEncoderOptions readsEncoderOptions;
+    final private String displayName;
+
+    /**
+     * Create a BAM encoder for the given output bundle. The primary resource in the bundle must
+     * have content type {@link BundleResourceType#ALIGNED_READS} (to find a decoder for a bundle,
+     * see {@link htsjdk.beta.plugin.registry.ReadsResolver}).
+     *
+     * @param outputBundle bundle to encode
+     * @param readsEncoderOptions options to use
+     */
+    public BAMEncoder(final Bundle outputBundle, final ReadsEncoderOptions readsEncoderOptions) {
```

Comment:
I like the idea in general, but it complicates things a bit in that it would require another type param on the encoders/decoders, and it will require adding a lot of boilerplate to each subclass (ReadsBundle, VariantsBundle, HapRefBundle, etc.). Currently ReadsBundle is just syntactic sugar for the super common case of reads and reads/index pairs that are IOPaths. If we require typed bundles that correspond to each content type, then each subclass will need to support every possible type of resource, such as InputStream, OutputStream, SeekableStream, etc.